src/pyhf/constraints.py

```diff
@@ -38,85 +44,105 @@ def _precompute(self):
             )
-            normal_sigmas = default_backend.concatenate(
-                list(map(default_backend.astensor, normal_constraint_sigmas))
+
+            _normal_sigmas = default_backend.concatenate(normal_constraint_sigmas)
+            sigmas = default_backend.reshape(_normal_sigmas, (1, -1))  # (1, normals)
+            self._batched_sigmas = default_backend.tile(
+                sigmas, (self.batch_size or 1, 1)
             )
-            normal_data = default_backend.concatenate(
-                list(
-                    map(
-                        lambda x: default_backend.astensor(x, dtype='int'),
-                        normal_constraint_data,
-                    )
+
+            # self.parameter_helper.index_selection  is [ (batch, parslice) ] ]
+            access_field = None
+            for x in self.parameter_helper.index_selection:
+                access_field = (
+                    x
+                    if access_field is None
+                    else default_backend.concatenate([access_field, x], axis=1)
                 )
-            )
+            # access field is (nbatch, normals)
+            self._access_field = access_field
 
-            self.normal_data = tensorlib.astensor(
-                default_backend.tolist(normal_data), dtype='int'
-            )
-            self.normal_sigmas = tensorlib.astensor(
-                default_backend.tolist(normal_sigmas)
-            )
-            self.normal_mean_idc = tensorlib.astensor(
-                default_backend.tolist(normal_mean_idc), dtype='int'
-            )
         else:
-            self.normal_data, self.normal_sigmas, self.normal_mean_idc = (
-                None,
-                None,
-                None,
-            )
+            self._normal_data = None
+            self._batched_sigmas = None
+            self._access_field = None
+
+        self._precompute()
+        events.subscribe('tensorlib_changed')(self._precompute)
+
+    def _precompute(self):
+        if not self.parameter_helper.index_selection:
+            return
+        tensorlib, _ = get_backend()
+        self.batched_sigmas = tensorlib.astensor(self._batched_sigmas)
+        self.normal_data = tensorlib.astensor(self._normal_data, dtype='int')
+        self.access_field = tensorlib.astensor(self._access_field, dtype='int')
 
     def logpdf(self, auxdata, pars):
-        if self.normal_data is None:
-            return 0
         tensorlib, _ = get_backend()
+        if not self.parameter_helper.index_selection:
+            return (
+                tensorlib.zeros(self.batch_size)
+                if self.batch_size is not None
+                else tensorlib.astensor(0.0)[0]
+            )
+
+        pars = tensorlib.astensor(pars)
+        if self.batch_size == 1 or self.batch_size is None:
+            batched_pars = tensorlib.reshape(
+                pars, (self.batch_size or 1,) + tensorlib.shape(pars)
```

Comment:
I don't understand this piece of code...

```
>>> import numpy as np
>>> pars = np.random.random(100)
>>> batch_size = 10
>>> np.reshape(pars, (batch_size or 1,) + np.shape(pars))
ValueError: cannot reshape array of size 100 into shape (10,100)
```

Shouldn't it be

```
tensorlib.reshape(pars, (self.batch_size or 1, -1))
```

Reply:
in your example batch size is "1"

```
import numpy as np
pars = np.random.random(100)
batch_size = 1
np.reshape(pars, (batch_size or 1,) + np.shape(pars))
```

Reply:
this code reshapes the parameters to be amenable to batched computation. If unbatched the (Npars,) tensor is reshaped into `(1,Npars)` .. in the batched case this is more or less a noop

Reply:
Ok, I would like to have a comment in the code here along that effect

```suggestion
                # if batched, noop
                # if unbatched, reshape parameters to (1, N parameters)
```